tracing-subscriber/src/fmt/fmt_layer.rs

```diff
@@ -0,0 +1,305 @@
+                Ok(buf) => {
+                    a = buf;
+                    &mut *a
+                }
+                _ => {
+                    b = String::new();
+                    &mut b
+                }
+            };
+
+            let ctx = self.make_ctx(ctx);
+            if self.fmt_event.format_event(&ctx, &mut buf, event).is_ok() {
+                let mut writer = self.make_writer.make_writer();
+                let _ = io::Write::write_all(&mut writer, buf.as_bytes());
+            }
+
+            buf.clear();
+        });
+    }
+}
+
+/// Provides the current span context to a formatter.
+pub struct FmtContext<'a, S, N> {
+    pub(crate) ctx: Context<'a, S>,
+    pub(crate) fmt_fields: &'a N,
+}
+
+impl<'a, S, N> fmt::Debug for FmtContext<'a, S, N> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FmtContext").finish()
+    }
+}
+
+impl<'a, S, N> FormatFields<'a> for FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    fn format_fields<R: RecordFields>(
+        &self,
+        writer: &'a mut dyn fmt::Write,
+        fields: R,
+    ) -> fmt::Result {
+        self.fmt_fields.format_fields(writer, fields)
+    }
+}
+
+impl<'a, S, N> FmtContext<'a, S, N>
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup> + LookupMetadata,
+    N: for<'writer> FormatFields<'writer> + 'static,
+{
+    /// Visits parent spans. Used to visit parent spans when formatting spans
+    /// and events
+    pub fn visit_spans<E, F>(&self, mut f: F) -> Result<(), E>
+    where
+        F: FnMut(&SpanRef<'_, S>) -> Result<(), E>,
+    {
+        let current_span = self.ctx.current_span();
+        let id = match current_span.id() {
+            Some(id) => id,
+            None => return Ok(()),
+        };
+        let span = match self.ctx.span(id) {
+            Some(span) => span,
+            None => return Ok(()),
+        };
+
+        #[cfg(feature = "smallvec")]
+        type SpanRefVec<'span, S> = smallvec::SmallVec<[SpanRef<'span, S>; 16]>;
+        #[cfg(not(feature = "smallvec"))]
+        type SpanRefVec<'span, S> = Vec<SpanRef<'span, S>>;
+
+        // an alternative way to handle this would be to the recursive approach that
+        // `fmt` uses that _does not_ entail any allocation in this fmt'ing
+        // spans path. however, that requires passing the store to `visit_spans`
+        // with a different lifetime, and i'm too lazy to sort that out now. this
+        // workaround shouldn't remaining in the final shipping version _unless_
+        // benchmarks show that small-vector optimization is preferable to not-very-deep
+        // recursion.
+        let parents = span.parents().collect::<SpanRefVec<'_, _>>();
+        let mut iter = parents.iter().rev();
+        // visit all the parent spans...
+        while let Some(parent) = iter.next() {
```

Comment:
nit: it might be more idiomatic to say 
```suggestion
        for parent in iter {
```
or even
```suggestion
        for parent in parents.iter().rev() {
```
and get rid of the `let iter = `